File: Packet/Hit.cs
using System.IO;
using Resources.Utilities;

namespace Resources.Packet {
    public class Hit : Packet {
        public long attacker;
        public long target;
        public float damage;
        public bool critical;//bool?
        //3pad
        public int stuntime;
        public int paddingA;
        public LongVector position = new LongVector();
        public FloatVector direction = new FloatVector();
        public byte isYellow;
        public DamageType type;
        public bool showlight;
        public byte paddingB;

        public Hit() : base() {
            PacketID = PacketID.hit;
        }
        public Hit(BinaryReader reader) : this() {
            attacker = reader.ReadInt64();
            target = reader.ReadInt64();
            damage = reader.ReadSingle();
            critical = reader.ReadBoolean();
            reader.ReadBytes(3);
            stuntime = reader.ReadInt32();
            paddingA = reader.ReadInt32();
            position = new LongVector(reader);
            direction = new FloatVector(reader);
            isYellow = reader.ReadByte();
            type = (DamageType)reader.ReadByte();
            showlight = reader.ReadBoolean();
            paddingB = reader.ReadByte();
        }

        protected override void WritePacketData(BinaryWriter writer) {
            writer.Write(attacker);
            writer.Write(target);
            writer.Write(damage);
            writer.Write(critical);
            writer.Write(new byte[3]);
            writer.Write(stuntime);
            writer.Write(paddingA);
            position.Write(writer);
            direction.Write(writer);
            writer.Write(isYellow);
            writer.Write((byte)type);
            writer.Write(showlight);
            writer.Write(paddingB);
        }
    }
}

File: src/main/kotlin/packets/Hit.kt
package com.github.lastexceed.cubeworldnetworking.packets

import com.github.lastexceed.cubeworldnetworking.utils.*

data class Hit(
	val attacker: CreatureId,
	val target: CreatureId,
	val damage: Float,
	val critical: Boolean,
	val paddingA: Byte = 0,
	val paddingB: Short = 0,
	val stuntime: Int,
	val paddingC: Int = 0,//todo: cuwo says this is something
	val position: Vector3<Long>,
	val direction: Vector3<Float>,
	val isYellow: Boolean,
	val type: Type,
	val flash: Boolean,
	val paddingD: Byte = 0
) : Packet(PacketId.Hit), WorldUpdate.SubPacket {
	override suspend fun writeTo(writer: Writer) {
		attacker.writeTo(writer)
		target.writeTo(writer)
		writer.writeFloat(damage)
		writer.writeBoolean(critical)
		writer.writeByte(paddingA)
		writer.writeShort(paddingB)
		writer.writeInt(stuntime)
		writer.writeInt(paddingC)
		writer.writeVector3Long(position)
		writer.writeVector3Float(direction)
		writer.writeBoolean(isYellow)
		type.writeTo(writer)
		writer.writeBoolean(flash)
		writer.writeByte(paddingD)
	}

	companion object : CwDeserializer<Hit> {
		override suspend fun readFrom(reader: Reader) =
			Hit(
				attacker = CreatureId.readFrom(reader),
				target = CreatureId.readFrom(reader),
				damage = reader.readFloat(),
				critical = reader.readBoolean(),
				paddingA = reader.readByte(),
				paddingB = reader.readShort(),
				stuntime = reader.readInt(),
				paddingC = reader.readInt(),
				position = reader.readVector3Long(),
				direction = reader.readVector3Float(),
				isYellow = reader.readBoolean(),
				type = Type.readFrom(reader),
				flash = reader.readBoolean(),
				paddingD = reader.readByte()
			)
	}

	enum class Type : CwSerializableEnumByte {
		Default,
		Block,
		Unknown,
		Miss,
		Dodge,
		Absorb,
		Invisible;

		companion object : CwEnumByteDeserializer<Type> {
			override val values = values()
		}
	}
}